dd-sdk-android/src/main/kotlin/com/datadog/android/sessionreplay/internal/domain/SessionReplayRequestFactory.kt

```diff
@@ -6,41 +6,178 @@
 
 package com.datadog.android.sessionreplay.internal.domain
 
-import com.datadog.android.core.internal.net.DataOkHttpUploaderV2
-import com.datadog.android.sessionreplay.internal.net.SessionReplayOkHttpUploader
+import com.datadog.android.core.internal.utils.sdkLogger
+import com.datadog.android.log.Logger
+import com.datadog.android.rum.RumAttributes
 import com.datadog.android.sessionreplay.net.BatchesToSegmentsMapper
 import com.datadog.android.v2.api.Request
 import com.datadog.android.v2.api.RequestFactory
 import com.datadog.android.v2.api.context.DatadogContext
+import okhttp3.RequestBody
+import okio.Buffer
+import java.io.IOException
+import java.util.Locale
 import java.util.UUID
 
 internal class SessionReplayRequestFactory(
-    private val sessionReplayOkHttpUploader: SessionReplayOkHttpUploader,
-    private val batchToSegmentsMapper: BatchesToSegmentsMapper = BatchesToSegmentsMapper()
+    private val endpoint: String,
+    private val batchToSegmentsMapper: BatchesToSegmentsMapper = BatchesToSegmentsMapper(),
+    private val requestBodyFactory: RequestBodyFactory = RequestBodyFactory(),
+    internal val internalLogger: Logger = sdkLogger,
+    private val bodyToByteArrayFactory: (RequestBody) -> ByteArray = {
+        val buffer = Buffer()
+        // this is handled below in the wrapper function
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        it.writeTo(buffer)
+        @Suppress("UnsafeThirdPartyFunctionCall")
+        buffer.readByteArray()
+    }
 ) : RequestFactory {
     override fun create(
         context: DatadogContext,
         batchData: List<ByteArray>,
         batchMetadata: ByteArray?
     ): Request {
-        // TODO: RUMM-2547 Drop this code and return a list of requests instead when
-        // the feature will be available in SDK V2
-        // Also add the necessary unit tests once this is done.
-        batchToSegmentsMapper.map(batchData).forEach {
-            sessionReplayOkHttpUploader.upload(
-                context,
-                it.first,
-                (it.second.toString() + "\n").toByteArray()
-            )
+        val serializedSegmentPair = batchToSegmentsMapper.map(batchData)
+        if (serializedSegmentPair == null) {
+            // TODO: RUMM-2397 Add the proper logs here once the sdkLogger will be added
+            // the data in the batch is corrupted. We return an empty request just to make
+            // sure the batch will be deleted by the core.
+            return getEmptyRequest()
```

Comment:
should we maybe change `RequestFactory` signature to allow null return value? Otherwise this case looks like a hack, which will also trigger some telemetry.

Reply:
I think this will be nice if we can do so.

Reply:
Yes, empty request looks like a hack and we shouldn't even make an attempt to send it.